### PR TITLE
Code coverage tracking

### PR DESCRIPTION
**Pull request by arindamnayak**
Signed-off-by: Arindam Nayak <arindam.nayak@outlook.com>

Here are the changes I did.
- This will push coverage report and sonar analysis to sonarcloud - vitess project. 
- This will happen every time some code is pushed to vitessio master branch only.
- It executes all unit tests,  endtoend test (go/test/endtoend) gathers the coverage report and push those to sonar.
- To get the coverage from endtoend test (where actual binaries are executed as part of testcase), we need to have instrumented executable binary, which is produced temporarily by `all_test_for_coverage.sh` and during test run those coverage files are generated.
- In sonar-project.properties file, we exclude certain paths for coverage analysis.

Things need to make it work.
- To make this work, we need to register vitessio project in sonarcloud.io (currently I did from my forked repo for testing), then use the sonarcloud key in vitess repo--> setting -> secret key --> add new key with name `SONAR_TOKEN` .

### CODE REVIEW

File: go/test/endtoend/cluster/vtctl_process.go [hunk 1]
```diff
@@ -71,13 +79,16 @@ func (vtctl *VtctlProcess) ExecuteCommandWithOutput(args ...string) (result stri
...
+		args = append([]string{"-test.coverprofile=" + getCoveragePath("vtctl-o-"+args[0]+".out"), "-test.v"}, args...)
+	}
 	tmpProcess := exec.Command(
 		vtctl.Binary,
 		args...,
 	)
 	log.Info(fmt.Sprintf("Executing vtctlclient with arguments %v", strings.Join(tmpProcess.Args, " ")))
 	resultByte, err := tmpProcess.CombinedOutput()
-	return string(resultByte), err
+	return filterResultWhenRunsForCoverage(string(resultByte)), err
```
Comment (ajeetj) [inline go/test/endtoend/cluster/vtctl_process.go]: Nit:

As we are doing check for `isCoverage` already. Maybe we do the same for return results as well and only call filterResult method when isCoverage is true

File: tools/all_test_for_coverage.sh [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+# These test uses excutables and launch them as process
+# After that all tests run, here we are testing those
+
+# All Go packages with test files.
+# Output per line: <full Go package name> <all _test.go files in the package>*
+
```
Comment (ajeetj) [inline tools/all_test_for_coverage.sh]: Let's refine these comments.

File: go/test/endtoend/cluster/vtctl_process.go (hunk 1 above)
Comment (arindamnayak) [reply go/test/endtoend/cluster/vtctl_process.go]: I check for `isCoverage` inside that function(`filterResultWhenRunsForCoverage`), as that function is being used from other place as well. I did not add that condition here, as I don't want to duplicate that condition in multiple places.

**Comment by deepthi** [comment]
> * To make this work, we need to register vitessio project in sonarcloud.io (currently I did from my forked repo for testing), then use the sonarcloud key in vitess repo--> setting -> secret key --> add new key with name `SONAR_TOKEN` .

Is this still in progress, or can the PR be merged after review?

File: tools/all_test_for_coverage.sh (hunk 2 above)
Comment (arindamnayak) [reply tools/all_test_for_coverage.sh]: Done

**Comment by arindamnayak** [comment]
> > * To make this work, we need to register vitessio project in sonarcloud.io (currently I did from my forked repo for testing), then use the sonarcloud key in vitess repo--> setting -> secret key --> add new key with name `SONAR_TOKEN` .
> 
> Is this still in progress, or can the PR be merged after review?

I don't have write access in this repo to configure this and before we merge it, we need to complete the steps I mentioned.

**Comment by deepthi** [comment]
@morgo can you help with the required configuration?

**Comment by deepthi** [comment]
> @morgo can you help with the required configuration?

@morgo ping

**Pull request merged**